# Week 3: Functions

$f(x)=x^2$

$f(2)=2^2=4$

$f(-1)=(-1)^2=1$

In [1]:
def f(x):
    result = x ** 2
    return result

In [2]:
print(f(4))

16


In [5]:
f(-1)

1

$f(x)=2x^3$

$f(2)=16$

In [20]:
def my_func(x):
    result = 2 * (x ** 3)
    return result

my_func(2)

16

### ทำไมต้องเขียน Function
เพื่อให้เป็นระเบียบ

Reuse ได้

Input: 0, 1, Many \
Output: 0, 1, Many

In [23]:
# 1 input \ 1 return
def f(x):
    result = x ** 3
    return result

f(5)

125

In [ ]:
# 0 return
def square_of_x(x):
    result = x ** 2

print(square_of_x(2))

None


In [ ]:
# Many returns
def f(x):
    return x, 2 * x, 3 * x

f(2) # TUPLE!

(2, 4, 6)

In [26]:
# 0 Input
def greet():
    print("Hello")

# greet(1)
greet()

Hello


In [28]:
# Many Inputs
def calculate_bmi(height_cm, weight_kg):
    bmi = weight_kg / ((height_cm / 100) ** 2)
    return bmi

calculate_bmi(161, 51)

19.675166853130666

$\Delta = \frac{1}{2} \times b \times h$

$\circ = \pi \times r^2$

In [32]:
import math
def calculate_circle_area(r):
    area = math.pi * r * r
    perimeter = 2 * math.pi * r
    return area, perimeter

calculate_circle_area(7)

(153.93804002589985, 43.982297150257104)

Heron's formula

$\Delta = \sqrt{s(s-a)(s-b)(s-c)}$

$s=\frac{a+b+c}{2}$

In [33]:
def area_triangle(a, b, c):
    s = (a+b+c) / 2
    area = math.sqrt(s * (s-a) * (s-b) * (s-c))
    return area

area_triangle(3, 4, 5)

6.0

## g.c.d (Greatest common divisor) ห.ร.ม

$gcd(6,10) = 2$

$gcd(12,8)=4$

$gcd(16,40)=8$

$gcd(7,9)=1$

In [ ]:
def gcd(x, y):
    start = max(x, y)
    for num in range(start, 0, -1):
        if (x % num == 0) and (y % num == 0):
            return num

gcd(16,40)

8


In [49]:
def gcd_v2(x, y):
    start = min(x, y)
    for num in range(start, 0, -1):
        if (x % num == 0) and (y % num == 0):
            return num

gcd(16,40)

8

In [ ]:
def gcd_euclidean(x, y):
    while y != 0:
        x, y = y, x % y

    return x

In [66]:
import time
t1 = time.time()
print(math.gcd(245680320, 17893459330, 473857, 432746))
t2 = time.time()
print(t2-t1)

1
0.0002677440643310547


## Least Common Multiple (l.c.m.) ค.ร.น

$lcm(4, 5)=20$

$lcm(4, 6)=12$

In [9]:
def lcm_v1(x, y):
    for num in range(1, x * y + 1, 1):
        if (num % x == 0) and (num % y == 0):
            return num
        
def lcm_v2(x, y):
    a = min(x, y)
    for num in range(a, x * y + 1, a):
        if (num % x == 0) and (num % y == 0):
            return num

def lcm_v3(x, y):
    a = max(x, y)
    for num in range(a, x * y + 1, a):
        if (num % x == 0) and (num % y == 0):
            return num

import time
import math
t1 = time.time()
# lcm_v1(47570, 4326)
t2 = time.time()
lcm_v2(4757000, 4326)
t3 = time.time()
lcm_v3(4757000, 4326)
t4 = time.time()
math.lcm(4757000, 4326)
t5 = time.time()
print(t2-t1, t3-t2, t4-t3, t5-t4)

0.00010085105895996094 0.14756512641906738 0.0002601146697998047 5.5789947509765625e-05


$5.58 \times 10^{-5}=0.0000558 = 55.8 \mu s$

### lcm-gcd identity
$lcm(a,b) \times gcd(a,b)=a \times b$

$lcm(4,6)=12$

$gcd(4,6)=2$

$lcm(a,b)=\frac{a \times b}{gcd(a,b)}$

In [ ]:
def lcm_v4(x, y):
    result = x * y / gcd_euclidean(x, y)
    return result